File: src/parser.rs
use crate::schema::{AstNodeType, CodeNode, SemanticFacets, UnresolvedRelation};
use tree_sitter::{Node, Parser};

pub struct AstParser;

impl AstParser {
    pub fn parse_file(
        filepath: &str,
        content: &str,
    ) -> Result<(Vec<CodeNode>, Vec<UnresolvedRelation>), String> {
        let mut parser = Parser::new();
        let extension = std::path::Path::new(filepath)
            .extension()
            .and_then(|ext| ext.to_str())
            .unwrap_or("");

        let mut nodes = Vec::new();
        let mut relations = Vec::new();
        let file_node_id = filepath.to_string();

        match extension {
            "rs" => {
                parser
                    .set_language(&tree_sitter_rust::LANGUAGE.into())
                    .map_err(|e| format!("Failed to load Rust grammar: {:?}", e))?;
                let tree = parser
                    .parse(content, None)
                    .ok_or_else(|| "Failed to parse Rust content".to_string())?;

                let root = tree.root_node();
                nodes.push(CodeNode {
                    id: file_node_id.clone(),
                    filepath: filepath.to_string(),
                    node_type: AstNodeType::File,
                    start_byte: 0,
                    end_byte: content.len(),
                    ast_hash: Self::hash_normalized(root, content),
                    semantics: SemanticFacets::default(),
                    ai_summary: None,
                    raw_code: content.to_string(),
                    previous_code: None,
                    is_dirty: true,
                });

                Self::walk_rust_node(
                    root,
                    content,
                    filepath,
                    Some(&file_node_id),
                    None,
                    None,
                    &mut nodes,
                    &mut relations,
                );
            }
            "dart" => {
                parser
                    .set_language(&tree_sitter_dart::LANGUAGE.into())
                    .map_err(|e| format!("Failed to load Dart grammar: {:?}", e))?;
                let tree = parser
                    .parse(content, None)
                    .ok_or_else(|| "Failed to parse Dart content".to_string())?;

                let root = tree.root_node();
                nodes.push(CodeNode {
                    id: file_node_id.clone(),
                    filepath: filepath.to_string(),
                    node_type: AstNodeType::File,
                    start_byte: 0,
                    end_byte: content.len(),
                    ast_hash: Self::hash_normalized(root, content),
                    semantics: SemanticFacets::default(),
                    ai_summary: None,
                    raw_code: content.to_string(),
                    previous_code: None,
                    is_dirty: true,
                });

                Self::walk_dart_node(
                    root,
                    content,
                    filepath,
                    Some(&file_node_id),
                    None,
                    None,
                    &mut nodes,
                    &mut relations,
                );
            }
            _ => return Err(format!("Unsupported file: .{}", extension)),
        }

        Ok((nodes, relations))
    }

    /// Strips comments and all semantic whitespace by walking named AST nodes [2]
    pub fn hash_normalized(node: Node, content: &str) -> String {
        let mut hasher = blake3::Hasher::new();
        let mut cursor = node.walk();
        
        let mut visit_stack = vec![node];
        while let Some(current) = visit_stack.pop() {
            if current.is_named() && current.kind() != "comment" {
                let start = current.start_byte();
                let end = current.end_byte();
                if end <= content.len() {
                    hasher.update(content[start..end].as_bytes());
                }
            }
            
            let mut children: Vec<Node> = current.children(&mut cursor).collect();
            children.reverse();
            for child in children {
                visit_stack.push(child);
            }
        }

        hasher.finalize().to_hex().to_string()
    }

    fn get_node_text<'a>(node: Node, content: &'a str) -> &'a str {
        let start = node.start_byte();
        let end = node.end_byte();
        if end <= content.len() {
            &content[start..end]
        } else {
            ""
        }
    }

    fn find_child_by_kind<'a>(node: Node<'a>, kind: &str) -> Option<Node<'a>> {
        let mut cursor = node.walk();
        for child in node.children(&mut cursor) {
            if child.kind() == kind {
                return Some(child);
            }
        }
        None
    }

    fn walk_rust_node(
        node: Node,
        content: &str,
        filepath: &str,
        parent_id: Option<&str>,
        active_class_name: Option<&str>, // For scope resolution [1]
        active_function_id: Option<&str>,
        nodes: &mut Vec<CodeNode>,
        relations: &mut Vec<UnresolvedRelation>,
    ) {
        let kind = node.kind();
        let mut current_parent = parent_id.map(String::from);
        let mut current_class = active_class_name.map(String::from);
        let mut current_function = active_function_id.map(String::from);

        match kind {
            "struct_item" | "enum_item" | "trait_item" => {
                let node_type = match kind {
                    "struct_item" => AstNodeType::Struct,
                    "enum_item" => AstNodeType::Enum,
                    _ => AstNodeType::Trait,
                };

                if let Some(ident_node) = Self::find_child_by_kind(node, "type_identifier") {
                    let name = Self::get_node_text(ident_node, content);
                    let id = format!("{}::{}", filepath, name);
                    current_parent = Some(id.clone());
                    current_class = Some(name.to_string());

                    nodes.push(Self::build_code_node(
                        id.clone(),
                        filepath,
                        node_type,
                        node,
                        content,
                    ));

                    if let Some(p) = parent_id {
                        relations.push(UnresolvedRelation::Contains {
                            source_id: p.to_string(),
                            target_id: id,
                        });
                    }
                }
            }
            "impl_item" => {
                let trait_name = Self::find_child_by_kind(node, "type_identifier")
                    .map(|n| Self::get_node_text(n, content).to_string());

                let struct_name = if trait_name.is_some() {
                    let mut cursor = node.walk();
                    let child_node = node.children(&mut cursor).find(|child| {
                        child.kind() == "type_identifier"
                            && Some(*child) != Self::find_child_by_kind(node, "type_identifier")
                    });

                    child_node.map(|s| Self::get_node_text(s, content).to_string())
                } else {
                    Self::find_child_by_kind(node, "type_identifier")
                        .map(|s| Self::get_node_text(s, content).to_string())
                };

                if let Some(ref s_name) = struct_name {
                    current_class = Some(s_name.clone());
                    let struct_id = format!("{}::{}", filepath, s_name);

                    if let Some(ref t_name) = trait_name {
                        relations.push(UnresolvedRelation::Implements {
                            source_id: struct_id,
                            target_symbol: t_name.clone(),
                        });
                    }
                }

                let target_name = trait_name
                    .or(struct_name)
                    .unwrap_or_else(|| "Anonymous".to_string());
                current_parent = Some(format!("{}::impl_{}", filepath, target_name));
            }
            "function_item" => {
                if let Some(ident_node) = Self::find_child_by_kind(node, "identifier") {
                    let name = Self::get_node_text(ident_node, content);
                    let (id, node_type) = match &parent_id {
                        Some(p_id) => (format!("{}::{}", p_id, name), AstNodeType::Method),
                        None => (format!("{}::{}", filepath, name), AstNodeType::Function),
                    };

                    nodes.push(Self::build_code_node(
                        id.clone(),
                        filepath,
                        node_type,
                        node,
                        content,
                    ));
                    current_function = Some(id.clone());

                    if let Some(p) = parent_id {
                        relations.push(UnresolvedRelation::Contains {
                            source_id: p.to_string(),
                            target_id: id,
                        });
                    }
                }
            }
            "call_expression" | "scoped_identifier" => {
                let called_symbol = if let Some(func_node) = Self::find_child_by_kind(node, "identifier") {
                    Self::get_node_text(func_node, content).to_string()
                } else if let Some(ident_node) = Self::find_child_by_kind(node, "field_identifier") {
                    Self::get_node_text(ident_node, content).to_string()
                } else {
                    String::new()
                };
                if !called_symbol.is_empty() {
                    if let Some(ref active_fn) = current_function {
                        relations.push(UnresolvedRelation::Calls {
                            source_id: active_fn.clone(),
                            target_symbol: called_symbol,
                            caller_filepath: filepath.to_string(),
                            caller_class_symbol: current_class.clone(),
                        });
                    }
                }
            }
            "method_call_expression" => {
                if let Some(method_node) = Self::find_child_by_kind(node, "field_identifier") {
                    let called_symbol = Self::get_node_text(method_node, content);
                    if let Some(ref active_fn) = current_function {
                        relations.push(UnresolvedRelation::Calls {
                            source_id: active_fn.clone(),
                            target_symbol: called_symbol.to_string(),
                            caller_filepath: filepath.to_string(),
                            caller_class_symbol: current_class.clone(),
                        });
                    }
                }
            }
            _ => {}
        }

        let mut cursor = node.walk();
        for child in node.children(&mut cursor) {
            Self::walk_rust_node(
                child,
                content,
                filepath,
                current_parent.as_deref(),
                current_class.as_deref(),
                current_function.as_deref(),
                nodes,
                relations,
            );
        }
    }

    fn walk_dart_node(
        node: Node,
        content: &str,
        filepath: &str,
        parent_id: Option<&str>,
        active_class_name: Option<&str>, // For scope resolution [1]
        active_function_id: Option<&str>,
        nodes: &mut Vec<CodeNode>,
        relations: &mut Vec<UnresolvedRelation>,
    ) {
        let kind = node.kind();
        let mut current_parent = parent_id.map(String::from);
        let mut current_class = active_class_name.map(String::from);
        let mut current_function = active_function_id.map(String::from);

        match kind {
            "class_declaration"
            | "mixin_declaration"
            | "enum_declaration"
            | "extension_declaration" => {
                // Support Enums & Extensions [3]
                let node_type = match kind {
                    "class_declaration" => AstNodeType::Class,
                    "mixin_declaration" => AstNodeType::Class,
                    "enum_declaration" => AstNodeType::Enum,
                    _ => AstNodeType::Extension,
                };

                if let Some(ident_node) = Self::find_child_by_kind(node, "type_identifier") {
                    let name = Self::get_node_text(ident_node, content);
                    let id = format!("{}::{}", filepath, name);
                    current_parent = Some(id.clone());
                    current_class = Some(name.to_string());

                    nodes.push(Self::build_code_node(
                        id.clone(),
                        filepath,
                        node_type,
                        node,
                        content,
                    ));

                    if let Some(p) = parent_id {
                        relations.push(UnresolvedRelation::Contains {
                            source_id: p.to_string(),
                            target_id: id.clone(),
                        });
                    }

                    let mut cursor = node.walk();
                    for child in node.children(&mut cursor) {
                        if child.kind() == "implements_clause" || child.kind() == "superclass" {
                            if let Some(type_node) =
                                Self::find_child_by_kind(child, "type_identifier")
                            {
                                let interface_symbol = Self::get_node_text(type_node, content);
                                relations.push(UnresolvedRelation::Implements {
                                    source_id: id.clone(),
                                    target_symbol: interface_symbol.to_string(),
                                });
                            }
                        }
                    }
                }
            }
            "method_declaration" => {
                if let Some(ident_node) = Self::find_child_by_kind(node, "identifier") {
                    let name = Self::get_node_text(ident_node, content);
                    let id = match &parent_id {
                        Some(p_id) => format!("{}::{}", p_id, name),
                        None => format!("{}::{}", filepath, name),
                    };

                    nodes.push(Self::build_code_node(
                        id.clone(),
                        filepath,
                        AstNodeType::Method,
                        node,
                        content,
                    ));
                    current_function = Some(id.clone());

                    if let Some(p) = parent_id {
                        relations.push(UnresolvedRelation::Contains {
                            source_id: p.to_string(),
                            target_id: id,
                        });
                    }
                }
            }
            "function_declaration" => {
                if let Some(ident_node) = Self::find_child_by_kind(node, "identifier") {
                    let name = Self::get_node_text(ident_node, content);
                    let id = format!("{}::{}", filepath, name);

                    nodes.push(Self::build_code_node(
                        id.clone(),
                        filepath,
                        AstNodeType::Function,
                        node,
                        content,
                    ));
                    current_function = Some(id.clone());

                    if let Some(p) = parent_id {
                        relations.push(UnresolvedRelation::Contains {
                            source_id: p.to_string(),
                            target_id: id,
                        });
                    }
                }
            }
            "method_invocation" => {
                if let Some(name_node) = Self::find_child_by_kind(node, "identifier") {
                    let called_symbol = Self::get_node_text(name_node, content);
                    if let Some(ref active_fn) = current_function {
                        relations.push(UnresolvedRelation::Calls {
                            source_id: active_fn.clone(),
                            target_symbol: called_symbol.to_string(),
                            caller_filepath: filepath.to_string(),
                            caller_class_symbol: current_class.clone(),
                        });
                    }
                }
            }
            _ => {}
        }

        let mut cursor = node.walk();
        for child in node.children(&mut cursor) {
            Self::walk_dart_node(
                child,
                content,
                filepath,
                current_parent.as_deref(),
                current_class.as_deref(),
                current_function.as_deref(),
                nodes,
                relations,
            );
        }
    }

    fn build_code_node(
        id: String,
        filepath: &str,
        node_type: AstNodeType,
        node: Node,
        content: &str,
    ) -> CodeNode {
        let start_byte = node.start_byte();
        let end_byte = node.end_byte();
        let code_slice = &content[start_byte..end_byte];

        let hash = Self::hash_normalized(node, content); // Run comments-and-whitespace normalization hash [2]

        CodeNode {
            id,
            filepath: filepath.to_string(),
            node_type,
            start_byte,
            end_byte,
            ast_hash: hash,
            semantics: SemanticFacets::default(),
            ai_summary: None,
            raw_code: code_slice.to_string(),
            previous_code: None,
            is_dirty: true,
        }
    }
}

File: src/parser.py
import os
from abc import ABC, abstractmethod
from pathlib import Path
from typing import List, Tuple, Optional, Dict
import blake3
from tree_sitter import Language, Parser, Query, Node as TsNode

from schema import (AstNodeType, CodeNode, SemanticFacets, UnresolvedRelation,
                    ContainsRelation, CallsRelation, ImplementsRelation,
                    FfiBridgeRelation, FfiExportRelation)

# Initialize Tree-Sitter Language Grammars
RUST_LANG = None
DART_LANG = None

try:
    import tree_sitter_rust as ts_rust
    RUST_LANG = Language(ts_rust.language())
except ImportError:
    pass

try:
    import tree_sitter_dart as ts_dart
    DART_LANG = Language(ts_dart.language())
except ImportError:
    pass


class LanguageParser(ABC):
    """Abstract strategy for language-specific parser implementations."""

    @property
    @abstractmethod
    def extension(self) -> str:
        """The file extension handled by this parser."""
        pass

    @property
    @abstractmethod
    def query_name(self) -> str:
        """The query scm file name representing this language parser."""
        pass

    @abstractmethod
    def get_language(self) -> Optional[Language]:
        """Returns the tree-sitter Language instance."""
        pass

    @abstractmethod
    def get_node_type_for_class(self) -> AstNodeType:
        """Determines the semantic AST node type mapped to class definitions."""
        pass


class RustParser(LanguageParser):
    @property
    def extension(self) -> str:
        return "rs"

    @property
    def query_name(self) -> str:
        return "rust"

    def get_language(self) -> Optional[Language]:
        return RUST_LANG

    def get_node_type_for_class(self) -> AstNodeType:
        return AstNodeType.STRUCT


class DartParser(LanguageParser):
    @property
    def extension(self) -> str:
        return "dart"

    @property
    def query_name(self) -> str:
        return "dart"

    def get_language(self) -> Optional[Language]:
        return DART_LANG

    def get_node_type_for_class(self) -> AstNodeType:
        return AstNodeType.CLASS


PARSER_STRATEGIES: Dict[str, LanguageParser] = {
    "rs": RustParser(),
    "dart": DartParser(),
}


def handle_class_node(parser_cls, ts_node, filepath, file_node_id, content_bytes, strategy, nodes, relations, resolve_active_source):
    name_node = parser_cls._find_child_by_kind(ts_node, "type_identifier")
    if name_node:
        name = parser_cls._get_node_text(name_node, content_bytes)
        node_id = f"{filepath}::{name}"
        node_type = strategy.get_node_type_for_class()
        nodes.append(parser_cls._build_code_node(node_id, filepath, node_type, ts_node, content_bytes))
        relations.append(ContainsRelation(source_id=file_node_id, target_id=node_id))


def handle_function_node(parser_cls, ts_node, filepath, file_node_id, content_bytes, strategy, nodes, relations, resolve_active_source):
    name_node = parser_cls._find_child_by_kind(ts_node, "identifier")
    if name_node:
        name = parser_cls._get_node_text(name_node, content_bytes)
        node_id = f"{file_node_id}::{name}"
        nodes.append(parser_cls._build_code_node(node_id, filepath, AstNodeType.FUNCTION, ts_node, content_bytes))
        relations.append(ContainsRelation(source_id=file_node_id, target_id=node_id))


def handle_impl_target_node(parser_cls, ts_node, filepath, file_node_id, content_bytes, strategy, nodes, relations, resolve_active_source):
    name_node = parser_cls._find_child_by_kind(ts_node, "type_identifier")
    if name_node:
        name = parser_cls._get_node_text(name_node, content_bytes)
        node_id = f"{filepath}::impl_{name}"
        nodes.append(parser_cls._build_code_node(node_id, filepath, AstNodeType.STRUCT, ts_node, content_bytes))
        relations.append(ContainsRelation(source_id=file_node_id, target_id=node_id))


def handle_calls_relation(parser_cls, ts_node, filepath, file_node_id, content_bytes, strategy, nodes, relations, resolve_active_source):
    symbol = parser_cls._get_node_text(ts_node, content_bytes)
    relations.append(CallsRelation(source_id=resolve_active_source(ts_node), target_symbol=symbol, caller_filepath=filepath))


def handle_implements_relation(parser_cls, ts_node, filepath, file_node_id, content_bytes, strategy, nodes, relations, resolve_active_source):
    symbol = parser_cls._get_node_text(ts_node, content_bytes)
    relations.append(ImplementsRelation(source_id=resolve_active_source(ts_node), target_symbol=symbol))


def handle_ffi_bridge_relation(parser_cls, ts_node, filepath, file_node_id, content_bytes, strategy, nodes, relations, resolve_active_source):
    symbol = parser_cls._get_node_text(ts_node, content_bytes)
    relations.append(FfiBridgeRelation(source_id=resolve_active_source(ts_node), target_symbol=symbol, caller_filepath=filepath))


def handle_ffi_export_relation(parser_cls, ts_node, filepath, file_node_id, content_bytes, strategy, nodes, relations, resolve_active_source):
    symbol = parser_cls._get_node_text(ts_node, content_bytes)
    relations.append(FfiExportRelation(source_id=resolve_active_source(ts_node), target_symbol=symbol))


CAPTURE_HANDLERS = {
    "node.class": handle_class_node,
    "node.function": handle_function_node,
    "node.impl_target": handle_impl_target_node,
    "relation.calls": handle_calls_relation,
    "relation.implements": handle_implements_relation,
    "relation.ffi_bridge": handle_ffi_bridge_relation,
    "node.ffi_export": handle_ffi_export_relation,
}


class AstParser:

    @staticmethod
    def locate_query(lang: str) -> str:
        """Recursively resolves the location of Tree-Sitter SCM query files."""
        primary_path = Path(f"queries/{lang}.scm")
        if primary_path.exists():
            return primary_path.read_text(encoding="utf-8")

        current_dir = Path.cwd().resolve()
        for parent in [current_dir] + list(current_dir.parents):
            nested_query = parent / "queries" / f"{lang}.scm"
            if nested_query.exists():
                return nested_query.read_text(encoding="utf-8")

        raise FileNotFoundError(
            f"Critical: Query rules asset for '{lang}' could not be located in local topology."
        )

    @classmethod
    def parse_file(
            cls, filepath: str,
            content: str) -> Tuple[List[CodeNode], List[UnresolvedRelation]]:
        """Parses target source code and yields isolated structural nodes and logical edges."""
        suffix = Path(filepath).suffix.lstrip(".")
        strategy = PARSER_STRATEGIES.get(suffix)

        if not strategy:
            raise ValueError(f"Unsupported file extension: .{suffix}")

        lang = strategy.get_language()
        if lang is None:
            raise ValueError(f"{strategy.query_name.capitalize()} tree-sitter grammar is not installed.")

        query_src = cls.locate_query(strategy.query_name)
        return cls._exec_query(filepath, content, query_src, lang, strategy)

    @classmethod
    def _exec_query(
            cls, filepath: str, content: str, query_src: str,
            language: Language, strategy: LanguageParser
    ) -> Tuple[List[CodeNode], List[UnresolvedRelation]]:
        parser = Parser(language)
        # Tree-sitter expects source encoding bytes for correct index tracking
        content_bytes = content.encode("utf-8")
        tree = parser.parse(content_bytes)
        root = tree.root_node

        nodes: List[CodeNode] = []
        relations: List[UnresolvedRelation] = []
        file_node_id = filepath

        # Bootstrap structural file boundary node
        nodes.append(
            CodeNode(id=file_node_id,
                     filepath=filepath,
                     node_type=AstNodeType.FILE,
                     start_byte=0,
                     end_byte=len(content_bytes),
                     ast_hash=cls.hash_normalized(root, content_bytes),
                     semantics=SemanticFacets(),
                     raw_code=content,
                     is_dirty=True))

        query = Query(language, query_src)
        captures = query.captures(root)

        def resolve_active_source(ts_node: TsNode) -> str:
            """Traverses backward up the processed node matrix to locate scope ownership."""
            for n in reversed(nodes):
                if (ts_node.start_byte >= n.start_byte
                        and ts_node.end_byte <= n.end_byte and n.node_type
                        in (AstNodeType.FUNCTION, AstNodeType.METHOD,
                            AstNodeType.STRUCT, AstNodeType.CLASS)):
                    return n.id
            return file_node_id

        # Tree-sitter Python captures return a dictionary/list of tuples: (Node, capture_name)
        for ts_node, capture_name in captures:
            handler = CAPTURE_HANDLERS.get(capture_name)
            if handler:
                handler(cls, ts_node, filepath, file_node_id, content_bytes, strategy, nodes, relations, resolve_active_source)

        return nodes, relations

    @staticmethod
    def hash_normalized(root_node: TsNode, content_bytes: bytes) -> str:
        """
        Executes a deterministic non-whitespace token-level hash calculation
        over the node's AST layout to eliminate formatting variances.
        """
        hasher = blake3.blake3()
        visit_stack = [root_node]

        while visit_stack:
            current = visit_stack.pop()

            # Check for structural named children that aren't comments
            has_named_children = any(child.is_named and child.type != "comment"
                                     for child in current.children)

            if not has_named_children and current.is_named and current.type != "comment":
                start = current.start_byte
                end = current.end_byte

                if end <= len(content_bytes):
                    token_bytes = content_bytes[start:end]
                    # Filter structural whitespace characters out of the calculation pipeline
                    normalized_bytes = bytes([
                        b for b in token_bytes if b not in (32, 10, 13, 9)
                    ])  # ' ', '\n', '\r', '\t'
                    hasher.update(normalized_bytes)

            # Mirror Rust stack execution bounds by reversing child nodes traversal sequence
            for child in reversed(current.children):
                visit_stack.append(child)

        return hasher.hexdigest()

    @staticmethod
    def _get_node_text(ts_node: TsNode, content_bytes: bytes) -> str:
        start = ts_node.start_byte
        end = ts_node.end_byte
        if end <= len(content_bytes):
            return content_bytes[start:end].decode("utf-8", errors="ignore")
        return ""

    @staticmethod
    def _find_child_by_kind(ts_node: TsNode, kind: str) -> Optional[TsNode]:
        for child in ts_node.children:
            if child.type == kind:
                return child
        return None

    @classmethod
    def _build_code_node(cls, node_id: str, filepath: str,
                         node_type: AstNodeType, ts_node: TsNode,
                         content_bytes: bytes) -> CodeNode:
        start = ts_node.start_byte
        end = ts_node.end_byte
        raw_segment = content_bytes[start:end].decode("utf-8", errors="ignore")
        ast_hash = cls.hash_normalized(ts_node, content_bytes)

        return CodeNode(id=node_id,
                        filepath=filepath,
                        node_type=node_type,
                        start_byte=start,
                        end_byte=end,
                        ast_hash=ast_hash,
                        semantics=SemanticFacets(),
                        raw_code=raw_segment,
                        is_dirty=True)

